codec-dns/src/main/java/io/netty/handler/codec/dns/AbstractDnsMessage.java

```diff
@@ -0,0 +1,344 @@
+        questions = question;
+        return this;
+    }
+
+    @Override
+    public List<DnsRecord> answers() {
+        List<DnsRecord> answers = this.answers;
+        if (answers == null) {
+            this.answers = answers = new LinkedList<DnsRecord>();
+        }
+        return answers;
+    }
+
+    @Override
+    public List<DnsRecord> authorityRecords() {
+        List<DnsRecord> authorityRecords = this.authorityRecords;
+        if (authorityRecords == null) {
+            this.authorityRecords = authorityRecords = new LinkedList<DnsRecord>();
+        }
+        return authorityRecords;
+    }
+
+    @Override
+    public DnsRecord additionalRecord() {
+        final Object additionalRecords = this.additionalRecords;
+        if (additionalRecords == null) {
+            return null;
+        }
+        if (additionalRecords instanceof DnsRecord) {
+            return (DnsRecord) additionalRecords;
+        }
+
+        @SuppressWarnings("unchecked")
+        final LinkedList<DnsRecord> additionalRecordList = (LinkedList<DnsRecord>) additionalRecords;
+        if (additionalRecordList.isEmpty()) {
+            return null;
+        }
+
+        return additionalRecordList.getFirst();
+    }
+
+    @Override
+    public DnsMessage setAdditionalRecord(DnsRecord record) {
+        ObjectUtil.checkNotNull(record, "record");
+        additionalRecords = record;
+        return this;
+    }
+
+    @Override
+    public List<DnsRecord> additionalRecords() {
+        final Object additionalRecords = this.additionalRecords;
+        if (additionalRecords == null) {
+            List<DnsRecord> newList = new LinkedList<DnsRecord>();
+            this.additionalRecords = newList;
+            return newList;
+        }
+        if (additionalRecords instanceof DnsRecord) {
+            List<DnsRecord> newList = new LinkedList<DnsRecord>();
+            newList.add((DnsRecord) additionalRecords);
+            this.additionalRecords = newList;
+            return newList;
+        }
+
+        @SuppressWarnings("unchecked")
+        final LinkedList<DnsRecord> additionalRecordList = (LinkedList<DnsRecord>) additionalRecords;
+        return additionalRecordList;
+    }
+
+    @Override
+    public DnsMessage touch() {
+        return (DnsMessage) super.touch();
+    }
+
+    @Override
+    public DnsMessage touch(Object hint) {
+        if (leak != null) {
+            leak.record(hint);
+        }
+        return this;
+    }
+
+    @Override
+    public DnsMessage retain() {
+        return (DnsMessage) super.retain();
+    }
+
+    @Override
+    public DnsMessage retain(int increment) {
+        return (DnsMessage) super.retain(increment);
+    }
+
+    @Override
+    protected void deallocate() {
+        deallocate(questions);
+        deallocateList(answers);
+        deallocateList(authorityRecords);
+        deallocate(additionalRecords);
+    }
+
+    private static void deallocate(Object recordOrList) {
```

Comment:
`AbstractDnsMessage` is the only customer, so let me it here until there's another. It's probably too specialized?